### PR TITLE
Add a method signature check for `setup`

### LINKED ISSUE #12801: TypeError: setup() got an unexpected keyword argument 'stage'
## 🐛 Bug
Hello, while trying to run the `trainer.fit(model,data_module)` getting the following error:
```
[/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py](https://localhost:8080/#) in _call_setup_hook(self)
   1488 
   1489         if self.datamodule is not None:
-> 1490             self.datamodule.setup(stage=fn)
   1491         self._call_callback_hooks("setup", stage=fn)
   1492         self._call_lightning_module_hook("setup", stage=fn)
```
`
TypeError: setup() got an unexpected keyword argument 'stage'`

- PyTorch Lightning Version (e.g., 1.5.0): 1.5.0
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9): 3.7
- OS (e.g., Linux): Windows
- CUDA/cuDNN version: 
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.0.5
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

- Any other relevant information:

### Additional context

cc @carmocca @awaelchli @borda @ananthsub @ninginthecloud @jjenniferdai @rohitgr7

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12801

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @rohitgr7 @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/trainer/configuration_validator.py [hunk 1]
```diff
@@ -402,3 +403,9 @@ def _check_datamodule_checkpoint_hooks(trainer: "pl.Trainer") -> None:
             "`LightningDataModule.on_load_checkpoint` was deprecated in"
             " v1.6 and will be removed in v1.8. Use `load_state_dict` instead."
         )
+
+
+def _check_setup_method(model: "pl.LightningModule", datamodule: "pl.LightningDataModule"):
+    for obj in (model, datamodule):
+        if is_overridden("setup", obj) and not is_param_in_hook_signature(obj.setup, "stage"):
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:410]: Does this support `*args` and `**kwargs`?

File: pytorch_lightning/trainer/configuration_validator.py [hunk 2]
```diff
@@ -402,3 +403,9 @@ def _check_datamodule_checkpoint_hooks(trainer: "pl.Trainer") -> None:
             "`LightningDataModule.on_load_checkpoint` was deprecated in"
             " v1.6 and will be removed in v1.8. Use `load_state_dict` instead."
         )
+
+
+def _check_setup_method(model: "pl.LightningModule", datamodule: "pl.LightningDataModule"):
+    for obj in (model, datamodule):
```
Comment (carmocca) [inline pytorch_lightning/trainer/configuration_validator.py:409]: We should do it for Callbacks too

File: pytorch_lightning/trainer/configuration_validator.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/configuration_validator.py:410]: yes since explicit if False

**Review by Borda — APPROVED** [review]

**Review by otaj — APPROVED** [review]

File: pytorch_lightning/trainer/configuration_validator.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/configuration_validator.py:409]: callbacks have more arguments `pl_module`, `trainer`. Should we still do it?

Comment (carmocca) [reply pytorch_lightning/trainer/configuration_validator.py:409]: IMO yes.

Comment (rohitgr7) [reply pytorch_lightning/trainer/configuration_validator.py]: done!

**Review by akihironitta — APPROVED** [review]
LGTM! The change is user-facing, but no need for a changelog entry?

**Review by akihironitta — COMMENTED** [review]
@rohitgr7 Some of the existing tests are failing due to the change here. Could you have a look?

**Comment by rohitgr7** [comment]
@akihironitta fixed! mind review it again :)

File: tests/trainer/test_config_validator.py [hunk 3]
```diff
@@ -156,3 +158,38 @@ def test_trainer_manual_optimization_config(tmpdir):
...
+
+        with pytest.raises(MisconfigurationException, match="does not have a `stage` argument"):
+            trainer.fit(**kwargs)
+
+    trainer = Trainer(fast_dev_run=True, callbacks=[CustomBoringCallback()])
+    model = BoringModel()
+    dm = BoringDataModule()
+
+    with pytest.raises(MisconfigurationException, match="does not have a `stage` argument"):
+        trainer.fit(model, datamodule=dm)
```
Comment (carmocca) [inline tests/trainer/test_config_validator.py:195]: The dm is not necessary here

```suggestion

    with pytest.raises(MisconfigurationException, match="does not have a `stage` argument"):
        trainer.fit(model)
```

**Review by carmocca — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Pull request merged**